### PR TITLE
Move camera configurations close to camera classes

### PR DESCRIPTION
**Pull request by Xierumeng**
Improve locality of code (LOC) by moving the configuration classes close to the camera classes. The tradeoff is that users are required to import all camera classes for configurations. A possible mitigation is to create the configuration from a common data structure (e.g. dictionary from YAML).

### CODE REVIEW

**Comment by Xierumeng** [comment]
Before:
```
configurations
 ↑    ↑    ↑
 |    |  camera0, camera1, ...
 |    |  ↑
 |  factory
 |    ↑
user -+
```
```py
import camera_configurations
import camera_factory
...
if camera_selection == 0:
    config = camera_configurations.ConfigCamera0()
elif camera_selection == 1:
    config = camera_configurations.ConfigCamera1()
elif ...
...
camera_device = camera_factory(..., config)
```
After:
```
camera0, camera1
 ↑    ↑
 |  factory
 |    ↑
user -+
```
```py
import camera0
import camera1
...
import camera_factory
...
if camera_selection == 0:
    config = camera0.ConfigCamera0()
elif camera_selection == 1:
    config = camera1.ConfigCamera1()
elif ...
...
camera_device = camera_factory(..., config)
```

File: modules/camera/base_camera.py [hunk 1]
```diff
@@ -20,13 +18,14 @@ def create(
         cls,
         width: int,
         height: int,
-        config: camera_configurations.PiCameraConfig | camera_configurations.OpenCVCameraConfig,
+        config: object,
```
Comment (maxlou05) [inline modules/camera/base_camera.py:21]: So we're only specifiying type in the factory class?

File: modules/camera/camera_picamera2.py [hunk 2]
```diff
@@ -11,11 +11,54 @@
...
+        exposure_time: int = 250,
+        analogue_gain: float = 64.0,
+        contrast: float = 1.0,
+        maybe_lens_position: float | None = None,
+    ) -> None:
+        """
+        timeout: Getting image timeout in seconds.
+
+        exposure_time: Microseconds.
+        analogue_gain:
```
Comment (maxlou05) [inline modules/camera/camera_picamera2.py:33]: A part of ISO, analogueGain * DigitialGain * 100 = ISO. From 0 to suggested max is 16, but can go up to 64.

File: modules/camera/camera_picamera2.py [hunk 3]
```diff
@@ -11,11 +11,54 @@
...
+        """
+        camera_controls = {
+            "ExposureTime": self.exposure_time,
+            "AnalogueGain": self.analogue_gain,
+            "Contrast": self.contrast,
+        }
+
+        if self.maybe_lens_position is not None:
+            camera_controls["LensPosition"] = self.maybe_lens_position
+            camera_controls["AfMode"] = picamera2.controls.AfModeEnum.Manual
```
Comment (maxlou05) [inline modules/camera/camera_picamera2.py:56]: I think this needs an import: `from libcamera import controls`, then it's just `controls.AfModeEnum...`

File: modules/camera/camera_picamera2.py [hunk 4]
```diff
@@ -11,11 +11,54 @@
...
+        else:
+            camera_controls["LensPosition"] = 0.0
+            camera_controls["AfMode"] = picamera2.controls.AfModeEnum.Auto
```
Comment (maxlou05) [inline modules/camera/camera_picamera2.py:59]: Same with this one

File: modules/camera/camera_picamera2.py [hunk 5]
```diff
@@ -11,11 +11,54 @@
...
+        contrast: float = 1.0,
+        maybe_lens_position: float | None = None,
+    ) -> None:
+        """
+        timeout: Getting image timeout in seconds.
+
+        exposure_time: Microseconds.
+        analogue_gain:
+        contrast: 0.0 to 32.0 . 0.0 is no contrast, 1.0 is normal contrast, higher is more contrast.
+        lens_position: Position of the lens is dioptres (reciprocal of metres: 1/m ).
```
Comment (maxlou05) [inline modules/camera/camera_picamera2.py:35]: (0 means infinity)

**Review by maxlou05 — COMMENTED** [review]
Reviewed

File: modules/camera/camera_picamera2.py (hunk 2 above)
Comment (Xierumeng) [reply modules/camera/camera_picamera2.py:33]: I'm not sure I understand the 2nd part. The acceptable range of values is 0 to 64, but the suggested range to use is 0 to 16 ? Why is the default argument 64 then?

File: modules/camera/camera_picamera2.py (hunk 3 above)
Comment (Xierumeng) [reply modules/camera/camera_picamera2.py:56]: Done.

File: modules/camera/camera_picamera2.py (hunk 4 above)
Comment (Xierumeng) [reply modules/camera/camera_picamera2.py:59]: Done.

File: modules/camera/camera_picamera2.py (hunk 5 above)
Comment (Xierumeng) [reply modules/camera/camera_picamera2.py:35]: Done.

**Review by maxlou05 — APPROVED** [review]
Approved

**Pull request merged**